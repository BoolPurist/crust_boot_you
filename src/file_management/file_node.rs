//! TODO: Remove if used
#![allow(dead_code)]

use std::path::{Path, PathBuf};

use super::FileKind;

#[derive(Debug)]
pub struct FileNodeMeta {
    node_type: FileKind,
    source_path: PathBuf,
}

impl FileNodeMeta {
    pub fn new(node_type: FileKind, source_path: PathBuf) -> Self {
        Self {
            node_type,
            source_path,
        }
    }

    pub fn source_path(&self) -> &Path {
        self.source_path.as_path()
    }

    pub fn node_type(&self) -> &FileKind {
        &self.node_type
    }
}
