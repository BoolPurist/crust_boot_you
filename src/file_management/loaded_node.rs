use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
#[derive(Debug, Deserialize, Serialize, Clone)]
pub enum LoadedNode {
    Folder { path: PathBuf },
    File { path: PathBuf, content: String },
}

impl LoadedNode {
    pub fn prepend_root(self, root: &Path) -> Self {
        return match self {
            Self::File { path, content } => Self::File {
                path: create_new_p(root, &path),
                content,
            },
            Self::Folder { path } => Self::Folder {
                path: create_new_p(root, &path),
            },
        };

        fn create_new_p(left: &Path, right: &Path) -> PathBuf {
            left.join(right)
        }
    }
}
