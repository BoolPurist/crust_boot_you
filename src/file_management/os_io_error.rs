use std::io;

#[derive(Debug, Error)]
pub enum AppIoError {
    #[error("Io error: {0}.")]
    Io(#[source] io::Error),
    #[error("Io error: {0}.")]
    FsExtra(#[source] fs_extra::error::Error),
    #[error("Io error: path does not exits")]
    NotFound,
    #[error("Io error: {0}")]
    Custom(String),
}

impl AppIoError {
    pub fn custom(message: impl AsRef<str>) -> Self {
        let message = message.as_ref();
        Self::Custom(message.to_string())
    }
}
impl From<io::Error> for AppIoError {
    fn from(error: io::Error) -> Self {
        if error.kind() == io::ErrorKind::NotFound {
            Self::NotFound
        } else {
            Self::Io(error)
        }
    }
}

impl From<fs_extra::error::Error> for AppIoError {
    fn from(error: fs_extra::error::Error) -> Self {
        if let fs_extra::error::ErrorKind::NotFound = error.kind {
            Self::NotFound
        } else {
            Self::FsExtra(error)
        }
    }
}
