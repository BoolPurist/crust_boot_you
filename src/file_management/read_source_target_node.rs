use super::SourceTargetNode;

pub struct ReadTargetNode {
    source_target: SourceTargetNode,
    content: String,
}

pub struct AugementedeTargetNode(ReadTargetNode);
