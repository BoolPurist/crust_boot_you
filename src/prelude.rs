use std::io;

pub use crate::constants;
pub use anyhow::Context;

pub use derive_getters::Getters;
pub use derive_new::new;
pub type IoResult<T = ()> = Result<T, io::Error>;
pub type AppError = anyhow::Error;
pub type AppResult<T = ()> = Result<T, AppError>;

pub type ReturnToUser = AppResult<String>;
pub type PathResult = AppResult<PathBuf>;
pub type AppIoResult<T = ()> = Result<T, AppIoError>;

pub use crate::app_traits::file_manipulator::FileManipulator;
pub use crate::app_traits::path_provider::PathProvider;
pub use crate::app_traits::path_resolver::PathResolver;
pub use crate::file_management::AppIoError;
pub use crate::ValidTemplateName;
pub use std::path::{Path, PathBuf};

#[cfg(test)]
pub use testing::*;
#[cfg(test)]
pub mod testing {

    pub use serde::{Deserialize, Serialize};

    pub const TEST_INPUT_FOLDER_NAME: &str = "test_input";
    #[macro_export]
    macro_rules! from_ron_input_file {
        ($input:literal) => {{
            ron::from_str(include_str!(concat!("test_input/", $input)))
                .expect(concat!($input, "is not in a valid format"))
        }};
    }
    pub use from_ron_input_file;
}
