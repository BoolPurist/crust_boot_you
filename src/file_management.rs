use std::{fs::FileType, path::Path};

use crate::prelude::*;

pub use node_entry_meta::NodeEntryMeta;
pub use os_io_error::AppIoError;
pub use source_target_node::SourceTargetNode;

mod node_entry_meta;
mod os_io_error;
mod source_target_node;
pub mod write_transactions;

#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
#[cfg_attr(test, derive(Serialize, Deserialize, PartialOrd, Ord))]
pub enum FileKind {
    File,
    Folder,
    Symlink,
}

impl TryFrom<FileType> for FileKind {
    type Error = AppIoError;
    fn try_from(value: FileType) -> Result<Self, Self::Error> {
        let os_file_type = value;
        if os_file_type.is_dir() {
            Ok(FileKind::Folder)
        } else if os_file_type.is_file() {
            Ok(FileKind::File)
        } else if os_file_type.is_symlink() {
            Ok(FileKind::Symlink)
        } else {
            Err(AppIoError::Custom(format!(
                "File type {:?} is not supported",
                os_file_type
            )))
        }
    }
}

pub fn detect_file_kind(path: &Path) -> AppIoResult<FileKind> {
    let file_meta = std::fs::metadata(path)?;

    let os_file_type = file_meta.file_type();
    os_file_type.try_into()
}

pub fn construct_file_target_path(
    source_path: &Path,
    ensured_template_folder: &Path,
) -> PathResult {
    let file_name = source_path.file_name().ok_or(anyhow!(
        "Could not extract file name from {:?}",
        source_path
    ))?;

    let file_name = &Path::new(file_name);
    Ok(ensured_template_folder.join(file_name))
}
