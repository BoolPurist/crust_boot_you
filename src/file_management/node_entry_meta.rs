//! TODO: Remove if used
#![allow(dead_code)]

use crate::prelude::*;
use std::path::PathBuf;

use super::FileKind;

#[derive(Debug, Clone, new, Getters, PartialEq, Eq, Hash)]
#[cfg_attr(test, derive(Serialize, Deserialize, PartialOrd, Ord))]
pub struct NodeEntryMeta {
    node_type: FileKind,
    source_path: PathBuf,
}

impl From<NodeEntryMeta> for (PathBuf, FileKind) {
    fn from(value: NodeEntryMeta) -> Self {
        (value.source_path, value.node_type)
    }
}

impl<'a> From<&'a NodeEntryMeta> for (&'a Path, FileKind) {
    fn from(value: &'a NodeEntryMeta) -> Self {
        (&value.source_path, value.node_type)
    }
}
