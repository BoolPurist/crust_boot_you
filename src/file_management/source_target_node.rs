//! TODO: Remove if used
#![allow(dead_code)]

use crate::prelude::*;

use super::{FileKind, NodeEntryMeta};
use std::path::PathBuf;

#[derive(Debug)]
#[cfg_attr(test, derive(Serialize, Deserialize))]
pub struct SourceTargetNode {
    source_node: NodeEntryMeta,
    target: PathBuf,
}

impl SourceTargetNode {
    fn new(source_node: NodeEntryMeta, target: PathBuf) -> Self {
        Self {
            source_node,
            target,
        }
    }

    pub fn source_path(&self) -> &Path {
        self.source_node.source_path()
    }
    pub fn target_path(&self) -> &Path {
        self.target.as_path()
    }
    pub fn node_type(&self) -> FileKind {
        *self.source_node.node_type()
    }

    pub fn opt_many_from_many_sources(
        prefix_source: &Path,
        prefix_target: &Path,
        source_to_map: impl IntoIterator<Item = NodeEntryMeta>,
    ) -> Option<Vec<Self>> {
        source_to_map
            .into_iter()
            .map(|node| {
                let target_path = node
                    .source_path()
                    .strip_prefix(prefix_source)
                    .ok()
                    .map(|to_join_with| prefix_target.join(to_join_with))?;
                Some(Self::new(node, target_path))
            })
            .collect()
    }
}

#[cfg(test)]
mod testing {

    use super::*;

    #[test]
    fn test_some_if_valid_suffix_and_prefix() {
        let source = Path::new("/some/source");
        let target = Path::new("/some/target");

        let to_map: Vec<NodeEntryMeta> = from_ron_input_file!("source_paths_with_same_prefix.ron");

        // Act
        let actual = SourceTargetNode::opt_many_from_many_sources(source, target, to_map).unwrap();

        insta::assert_ron_snapshot!(actual);
    }
    #[test]
    fn test_none_if_one_file_with_wrong_prefix() {
        use crate::prelude::testing::*;
        let source = Path::new("/some/source");
        let target = Path::new("/some/target");

        let to_map: Vec<NodeEntryMeta> = from_ron_input_file!("different_prefixes.ron");

        // Act
        let actual = SourceTargetNode::opt_many_from_many_sources(source, target, to_map);

        assert!(actual.is_none());
    }
}
