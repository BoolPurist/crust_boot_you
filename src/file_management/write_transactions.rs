use std::path::PathBuf;

use derive_getters::Getters;
use derive_new::new;

use super::{FileKind, SourceTargetNode};

#[cfg(test)]
use crate::prelude::testing::*;

#[derive(Debug, Into)]
#[cfg_attr(test, derive(Serialize, Deserialize))]
pub struct WriteTransactions {
    files: Vec<FileToLoad>,
    folders: Vec<DirToEnsure>,
}

impl WriteTransactions {
    pub fn new(files_dirs: impl IntoIterator<Item = SourceTargetNode>) -> Self {
        let files_dirs = files_dirs.into_iter();
        let cap = files_dirs.size_hint().1.unwrap_or(0);
        let mut files = Vec::with_capacity(cap);
        let mut folders = Vec::with_capacity(cap);

        for next_file_or_dir in files_dirs {
            match next_file_or_dir.node_type() {
                FileKind::File => files.push(next_file_or_dir.into()),
                FileKind::Folder => folders.push(next_file_or_dir.into()),
                _ => unreachable!(),
            }
        }

        Self { files, folders }
    }
}

#[derive(Debug, Getters, Into, new)]
#[cfg_attr(test, derive(Serialize, Deserialize))]
pub struct DirToEnsure {
    source: PathBuf,
    target: PathBuf,
}

#[derive(Debug, Getters, Into, new)]
#[cfg_attr(test, derive(Serialize, Deserialize))]
pub struct FileToLoad {
    source: PathBuf,
    target: PathBuf,
}

impl From<SourceTargetNode> for FileToLoad {
    fn from(value: SourceTargetNode) -> Self {
        let (source, target) = value.into();
        FileToLoad::new(source, target)
    }
}
impl From<SourceTargetNode> for DirToEnsure {
    fn from(value: SourceTargetNode) -> Self {
        let (source, target) = value.into();
        DirToEnsure::new(source, target)
    }
}

#[cfg(test)]
mod testing {
    use crate::file_management::NodeEntryMeta;

    use super::*;

    #[test]
    fn separate_files_and_dirs() {
        let (source, target) = (PathBuf::from("/a/source"), PathBuf::from("/a/target"));
        let nodes = [
            NodeEntryMeta::new(FileKind::File, PathBuf::from("/a/source/ccc")),
            NodeEntryMeta::new(FileKind::File, PathBuf::from("/a/source/aaaa/aa")),
            NodeEntryMeta::new(FileKind::Folder, PathBuf::from("/a/source/aaaa")),
            NodeEntryMeta::new(FileKind::Folder, PathBuf::from("/a/source/aaaa/aaa")),
            NodeEntryMeta::new(FileKind::File, PathBuf::from("/a/source/zzz/z")),
            NodeEntryMeta::new(FileKind::Folder, PathBuf::from("/a/source/b")),
        ];
        let input = SourceTargetNode::opt_many_from_many_sources(&source, &target, nodes).unwrap();

        // Act
        let actual = WriteTransactions::new(input);

        insta::assert_ron_snapshot!(actual);
    }
}
