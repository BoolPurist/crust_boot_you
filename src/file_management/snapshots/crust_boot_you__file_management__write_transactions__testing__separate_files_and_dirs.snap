---
source: src/file_management/write_transactions.rs
expression: actual
---
WriteTransactions(
  files: [
    FileToLoad(
      source: "/a/source/ccc",
      target: "/a/target/ccc",
    ),
    FileToLoad(
      source: "/a/source/aaaa/aa",
      target: "/a/target/aaaa/aa",
    ),
    FileToLoad(
      source: "/a/source/zzz/z",
      target: "/a/target/zzz/z",
    ),
  ],
  folders: [
    DirToEnsure(
      source: "/a/source/aaaa",
      target: "/a/target/aaaa",
    ),
    DirToEnsure(
      source: "/a/source/aaaa/aaa",
      target: "/a/target/aaaa/aaa",
    ),
    DirToEnsure(
      source: "/a/source/b",
      target: "/a/target/b",
    ),
  ],
)
